test/strings.jl

```diff
@@ -1616,3 +1616,52 @@ d = UTF32String(c)
 c[1] = 'A'
 @test d=="A"
 
+# issue #11004 (#10959)
+
+function tstcvt(strUTF8::UTF8String, strUTF16::UTF16String, strUTF32::UTF32String)
+    @test utf16(strUTF8) == strUTF16
+    @test utf32(strUTF8) == strUTF32
+    @test utf8(strUTF16) == strUTF8
+    @test utf32(strUTF16) == strUTF32
+    @test utf8(strUTF32)  == strUTF8
+    @test utf16(strUTF32) == strUTF16
+end
+
+# Create some ASCII, UTF8, UTF16, and UTF32 strings
+strAscii = "abcdefgh"
+strA_UTF8 = ("abcdefgh\uff")[1:8]
+strL_UTF8 = "abcdef\uff\uff"
+str2_UTF8 = "abcd\uff\uff\u7ff\u7ff"
+str3_UTF8 = "abcd\uff\uff\u7fff\u7fff"
+str4_UTF8 = "abcd\uff\u7ff\u7fff\U7ffff"
+strS_UTF8 = UTF8String(b"abcd\xc3\xbf\xdf\xbf\xe7\xbf\xbf\xed\xa0\x80\xed\xb0\x80")
+strC_UTF8 = UTF8String(b"abcd\xc3\xbf\xdf\xbf\xe7\xbf\xbf\U10000")
+
+strA_UTF16 = utf16(strA_UTF8)
+strL_UTF16 = utf16(strL_UTF8)
+str2_UTF16 = utf16(str2_UTF8)
+str3_UTF16 = utf16(str3_UTF8)
+str4_UTF16 = utf16(str4_UTF8)
+strS_UTF16 = utf16(strS_UTF8)
+strA_UTF32 = utf32(strA_UTF8)
+strL_UTF32 = utf32(strL_UTF8)
+str2_UTF32 = utf32(str2_UTF8)
+str3_UTF32 = utf32(str3_UTF8)
+str4_UTF32 = utf32(str4_UTF8)
+strS_UTF32 = utf32(strS_UTF8)
+@test utf8(strAscii) == strAscii
+@test utf16(strAscii) == strAscii
+@test utf32(strAscii) == strAscii
+tstcvt(strA_UTF8,strA_UTF16,strA_UTF32)
+tstcvt(strL_UTF8,strL_UTF16,strL_UTF32)
+tstcvt(str2_UTF8,str2_UTF16,str2_UTF32)
+tstcvt(str3_UTF8,str3_UTF16,str3_UTF32)
+tstcvt(str4_UTF8,str4_UTF16,str4_UTF32)
+# need different tests where we are converting the surrogates or overlong \0
+@test utf16(strS_UTF8) == strC_UTF8
+@test utf32(strS_UTF8) == strC_UTF8
+@test utf8(strS_UTF16) == strC_UTF8
+@test utf32(strS_UTF16) == strC_UTF8
+@test utf8(strS_UTF32)  == strC_UTF8
+@test utf16(strS_UTF32) == strC_UTF8
```

Comment:
Yes, that would be good, I'll add some of those (of course, the current conversion code wouldn't pass, that is a lot of what this change is fixing)